### PR TITLE
Add crowdpose dataset

### PR DESCRIPTION
**Pull request by xiaohangcd**

### CODE REVIEW

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+                joints_3d_visible[ipt, 0] = t_vis
+                joints_3d_visible[ipt, 1] = t_vis
+                joints_3d_visible[ipt, 2] = 0
+
+            interference_num_joints = len(obj['keypoints']) // 3 - num_joints
+            interference_joints_3d = np.zeros((interference_num_joints, 3),
+                                              dtype=np.float)
+            interference_joints_3d_vis = np.zeros((interference_num_joints, 3),
+                                                  dtype=np.float)
+            for ipt in range(interference_num_joints):
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:187]: does the following code do the same thing as this loop?

```
interference_joints_3d = obj['keypoints'].reshape(-1, 3)[num_joints:num_joints+interference_num_joints, :]
interference_joints_3d_vis[...] = np.array([1, 1, 0]) 
```

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+                interference_joints_3d[
+                    ipt, 1] = obj['keypoints'][(num_joints + ipt) * 3 + 1]
+                interference_joints_3d[ipt, 2] = \
+                    obj['keypoints'][(num_joints + ipt) * 3 + 2]
+
+                interference_joints_3d_vis[ipt, 0] = 1
+                interference_joints_3d_vis[ipt, 1] = 1
+                interference_joints_3d_vis[ipt, 2] = 0
+
+            center, scale = self._box2cs(obj['clean_bbox'][:4])
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:199]: here

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py [hunk 3]
```diff
@@ -179,22 +179,8 @@ def _load_coco_keypoint_annotation_kernal(self, index):
...
-                    ipt, 0] = obj['keypoints'][(num_joints + ipt) * 3 + 0]
-                interference_joints_3d[
-                    ipt, 1] = obj['keypoints'][(num_joints + ipt) * 3 + 1]
-                interference_joints_3d[ipt, 2] = \
-                    obj['keypoints'][(num_joints + ipt) * 3 + 2]
-
-                interference_joints_3d_vis[ipt, 0] = 1
-                interference_joints_3d_vis[ipt, 1] = 1
-                interference_joints_3d_vis[ipt, 2] = 0
+            interference_joints_3d = np.array(obj['keypoints']).reshape(
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:182]: forgot the interference_joints_3d_vis?

Comment (xiaohangcd) [reply mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:182]: I deleted it because I found it unnecessary

File: README.md [hunk 4]
```diff
@@ -112,4 +112,4 @@ We appreciate all contributions to improve MMPose. Please refer to [CONTRIBUTING
 
 MMPose is an open source project that is contributed by researchers and engineers from various colleges and companies.
 We appreciate all the contributors who implement their methods or add new features, as well as users who give valuable feedbacks.
-We wish that the toolbox and benchmark could serve the growing research community by providing a flexible toolkit to reimplement existing methods and develop their own new models.
\ No newline at end of file
+We wish that the toolbox and benchmark could serve the growing research community by providing a flexible toolkit to reimplement existing methods and develop their own new models.
```
Comment (innerlee) [inline README.md:115]: what's the diff of this line?

Comment (xiaohangcd) [reply README.md:115]: I can't find and I think it changed by pre-commit

**Comment by innerlee** [comment]
is accuracy good?

File: mmpose/core/evaluation/top_down_eval.py [hunk 5]
```diff
@@ -272,3 +272,12 @@ def keypoints_from_heatmaps(heatmaps,
...
     return preds, maxvals
+
+
+def crowd_keypoints_from_heatmaps(heatmaps,
+                                  center,
+                                  scale,
+                                  post_process=True,
+                                  unbiased=False,
+                                  kernel=11):
+    pass
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:283]: evaluation will be added in separate pr, right?

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+        """Keypoint evaluation using CrowdPoseAPI."""
+        coco_dt = self.coco.loadRes(res_file)
+        coco_eval = COCOeval(self.coco, coco_dt, 'keypoints')
+        coco_eval.params.useSegm = None
+        coco_eval.evaluate()
+        coco_eval.accumulate()
+        coco_eval.summarize()
+
+        stats_names = [
+            'AP', 'Ap .5', 'AP .75', 'AP (M)', 'AP (L)', 'AR', 'AR .5',
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:389]: Ap

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+                'score': img_kpt['score'],
+                'center': list(img_kpt['center']),
+                'scale': list(img_kpt['scale'])
+            } for img_kpt, key_point in zip(img_kpts, key_points)]
+
+            cat_results.extend(result)
+
+        return cat_results
+
+    def _do_python_keypoint_eval(self, res_file):
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:379]: does "python" have special meaning?

File: mmpose/models/detectors/top_down_crowd.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+        if keypoint_head is not None:
+            self.keypoint_head = builder.build_head(keypoint_head)
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        self.loss = builder.build_loss(loss_pose)
+        self.init_weights(pretrained=pretrained)
+
+    @property
+    def with_keypoint(self):
```
Comment (innerlee) [inline mmpose/models/detectors/top_down_crowd.py:51]: avoid starting with `with`. `with` makes user think of the python with-statement.

maybe the name `has_keypoint` or `use_keypoint` or some other possible `xxx_keypoint` is better

File: mmpose/core/evaluation/top_down_eval.py (hunk 5 above)
Comment (xiaohangcd) [reply mmpose/core/evaluation/top_down_eval.py:283]: will be added in this pr

File: mmpose/models/detectors/top_down_crowd.py (hunk 8 above)
Comment (innerlee) [reply mmpose/models/detectors/top_down_crowd.py:51]: a second thought. with_xxx is not that bad.

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py (hunk 7 above)
Comment (xiaohangcd) [reply mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:379]: no. just named the same as  coco dataset's keypoint eval

Comment (innerlee) [reply mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:379]: okay we can give it a better name

File: mmpose/core/post_processing/post_transforms.py [hunk 9]
```diff
@@ -226,3 +226,55 @@ def rotate_point(pt, angle_rad):
     rotated_pt = [new_x, new_y]
 
     return rotated_pt
+
+
+def fliplr_interference_joints(joints, width, matched_parts):
+    """flip interference joint."""
```
Comment (innerlee) [inline mmpose/core/post_processing/post_transforms.py:232]: 1. what is an interference joint?
2. this is an inplace operation. do you need to protect the input by copy?

Comment (innerlee) [reply mmpose/core/post_processing/post_transforms.py:232]: image

Comment (xiaohangcd) [reply mmpose/core/post_processing/post_transforms.py:232]: a joint in someones bbox but not belong him

Comment (innerlee) [reply mmpose/core/post_processing/post_transforms.py:232]: it will be good to add this explanation to docstring

File: mmpose/core/post_processing/matching.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+
+            assert len(person[k]) > 0
+            x0, y0, s0 = person[k][0]
+            if s0 < 0.05:
+                continue
+            for g_id, g in kp_group.items():
+
+                x_c, y_c = kp_group[g_id]['group_center']
+                '''
+                Get Average Box Size
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:152]: Single line comment format is ok

File: mmpose/core/post_processing/matching.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+                '''
+                group_area = kp_group[g_id]['group_area']
+                group_area = group_area[0] * group_area[1] / (group_area[2]**2)
+                '''
+                Groupingn Criterion
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:157]: same here

**Comment by innerlee** [comment]
need to rebase master now

File: mmpose/core/post_processing/matching.py [hunk 12]
```diff
@@ -148,16 +148,13 @@ def _grouping(bboxes, pose_preds, num_joints, box_scores):
...
-                Get Average Box Size
-                '''
+                # Get Average Box Size
                 group_area = kp_group[g_id]['group_area']
                 group_area = group_area[0] * group_area[1] / (group_area[2]**2)
-                '''
-                Groupingn Criterion
-                '''
+
+                # Groupingn Criterion
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:155]: Groupingn  typo?

File: mmpose/core/evaluation/top_down_eval.py [hunk 13]
```diff
@@ -272,3 +273,112 @@ def keypoints_from_heatmaps(heatmaps,
...
+                                                 preds[n][k][m])
+        else:
+            # add +/-0.25 shift to the predicted locations for higher acc.
+            for n in range(N):
+                for k in range(K):
+                    heatmap = heatmaps[n][k]
+                    for m in range(M):
+                        px = int(preds[n][k][m][0])
+                        py = int(preds[n][k][m][1])
+                        if 1 < px < W - 1 and 1 < py < H - 1:
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:372]: `px > 1` or `> 0`?

Comment (xiaohangcd) [reply mmpose/core/evaluation/top_down_eval.py:372]: is right

File: mmpose/core/evaluation/top_down_eval.py [hunk 14]
```diff
@@ -272,3 +273,112 @@ def keypoints_from_heatmaps(heatmaps,
...
+    Returns:
+        np.ndarray[N, K, M, 2]: Predicted keypoint location.
+        np.ndarray[N, K, M, 1]: Scores (confidence) of the keypoints.
+    """
+    assert isinstance(heatmaps,
+                      torch.Tensor), ('heatmaps should be torch.Tensor')
+    assert heatmaps.dim() == 4, 'batch_images should be 4-ndim'
+
+    pooling = torch.nn.MaxPool2d(5, 1, 2)
+    maxm = pooling(heatmaps)
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:300]: `maxm = F.max_pool2d(heatmaps, 5, 1, 2)`

F is torch.nn.functional

File: mmpose/core/evaluation/top_down_eval.py (hunk 13 above)
Comment (innerlee) [reply mmpose/core/evaluation/top_down_eval.py:372]: why omit the case when px = 1 then

File: mmpose/core/post_processing/matching.py [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+        in_vis_thre (float): threshold.
+    Returns:
+        final_result (list): predicted in the image.
+    """
+
+    # Group same keypointns together
+    kp_groups = _grouping(bboxes, pose_preds, num_joints, box_scores)
+
+    # Generate Matrix
+    human_num = len(pose_preds.keys())
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:34]: does `len(pose_preds)` give the same answer?

Comment (innerlee) [reply mmpose/core/post_processing/matching.py:34]: about var naming: `num_human` is more consistent

Comment (innerlee) [reply mmpose/core/post_processing/matching.py:34]: in L16, there is a "num_person: N"

File: mmpose/core/post_processing/matching.py [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+    costMatrix = []
+
+    for k in range(num_joints):
+        kp_group = kp_groups[k]
+        joint_num = len(kp_group.keys())
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:39]: what's the difference between `num_joints` and `joint_num`

File: mmpose/core/post_processing/matching.py [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+        costMatrix.append(np.zeros((human_num, joint_num)))
+
+    for n, person in pose_preds.items():
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:42]: 1. what's the difference between person and human?
2. either use `n` or use `h_id` in this loop, not both of them

File: mmpose/core/post_processing/matching.py [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+    """
+
+    # Group same keypointns together
+    kp_groups = _grouping(bboxes, pose_preds, num_joints, box_scores)
+
+    # Generate Matrix
+    human_num = len(pose_preds.keys())
+    costMatrix = []
+
+    for k in range(num_joints):
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:37]:
```python
costMatrix = [np.zeros((human_num, len(kp_group)), dtype=np.float) for kp_group in kp_groups]
```

File: mmpose/core/post_processing/matching.py [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+        costMatrix.append(np.zeros((human_num, joint_num)))
+
+    for n, person in pose_preds.items():
+        h_id = n
+        assert 0 <= h_id < human_num
+        for k in range(num_joints):
+            g_id = person['group_id'][k]
+            if g_id is not None:
+
+                g_id = int(g_id) - 1
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:49]: add a comment on the `- 1`

File: mmpose/core/post_processing/matching.py [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+
+def _matching(pose_preds, matrix, num_joints, kp_groups):
+    """use hungarian algorithm to match person and keypoints.
+
+    Note:
+        num_person: N
+        num_keypoints: K
+    Args:
+        pose_preds (dict)
+        matric (list):cost matric.
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:217]: is matric a typo?

File: mmpose/datasets/pipelines/top_down_crowd_transform.py [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+            joints_3d: np.ndarray([num_joints, 3])
+            joints_3d_visible: np.ndarray([num_joints, 3])
+        Returns:
+             target, target_weight(1: visible, 0: invisible)
+        """
+        num_joints = cfg['num_joints']
+        image_size = cfg['image_size']
+        heatmap_size = cfg['heatmap_size']
+        joints_weight = cfg['joints_weight']
+        use_different_joints_weight = cfg['use_different_joints_weight']
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_crowd_transform.py:125]: use_different_joints_weight
this name has been changed

File: mmpose/core/post_processing/matching.py [hunk 22]
```diff
@@ -145,7 +142,7 @@ def _grouping(bboxes, pose_preds, num_joints, box_scores):
             x0, y0, s0 = person[k][0]
             if s0 < 0.05:
                 continue
-            for g_id, g in kp_group.items():
+            for g_id, _ in kp_group.items():
```
Comment (innerlee) [inline mmpose/core/post_processing/matching.py:145]: actually the `kp_group[g_id]` in this loop can be written as `g`

Comment (innerlee) [reply mmpose/core/post_processing/matching.py:145]: btw in the beginning of this function, 
`kp_groups = {k: {} for k in range(num_joints)}`

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 23]
```diff
@@ -298,8 +298,8 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
                     target[joint_id][img_y[0]:img_y[1], img_x[0]:img_x[1]] = \
                         g[g_y[0]:g_y[1], g_x[0]:g_x[1]]
 
-        if use_different_joint_weights:
-            target_weight = np.multiply(target_weight, joint_weights)
+        if use_different_joints_weight:
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py]: emmmm

Comment (xiaohangcd) [reply mmpose/datasets/pipelines/top_down_transform.py]: it's top_down_transform.py ,i modified top_down_crowd_transform.py. Should i modify these files now ? or you process when rebase master

Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py]: could you try to rebase master?

Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py]: okay, if you have problems in rebase, I will do it later

**Comment by innerlee** [comment]
@xiaohangcd I rebased this branch onto master and put it here https://github.com/open-mmlab/mmpose/pull/74
pls take a look
